src/lib/mina/transaction.ts

```diff
@@ -280,26 +306,74 @@ type RejectedTransaction = Pick<
   errors: string[];
 };
 
-type TransactionPromise = Promise<Transaction> & {
-  sign(...args: Parameters<Transaction['sign']>): TransactionPromise;
+/**
+ * A `Promise<Transaction>` with some additional methods for making chained method calls
+ * into the pending value upon its resolution.
+ */
+type TransactionPromise<
+  Proven extends boolean,
+  Signed extends boolean
+> = Promise<Transaction<Proven, Signed>> & {
+  /** Equivalent to calling the resolved `Transaction`'s `send` method. */
   send(): PendingTransactionPromise;
-};
+} & (Proven extends false
+    ? {
+        /**
+         * Calls `prove` upon resolution of the `Transaction`. Returns a
+         * new `TransactionPromise` with the field `proofPromise` containing
+         * a promise which resolves to the proof array.
+         */
+        prove(): TransactionPromise<true, Signed>;
+      }
+    : {
+        /**
+         * If the chain of method calls that produced the current `TransactionPromise`
+         * contains a `prove` call, then this field contains a promise resolving to the
+         * proof array which was output from the underlying `prove` call.
+         */
+        proofs: Promise<Transaction<true, Signed>['proofs']>;
+      }) &
+  (Signed extends false
+    ? {
+        /** Equivalent to calling the resolved `Transaction`'s `sign` method. */
+        sign(
+          ...args: Parameters<Transaction<Proven, Signed>['sign']>
+        ): TransactionPromise<Proven, true>;
+      }
+    : {});
 
-function toTransactionPromise(
-  getPromise: () => Promise<Transaction>
-): TransactionPromise {
+function toTransactionPromise<Proven extends boolean, Signed extends boolean>(
+  getPromise: () => Promise<Transaction<Proven, Signed>>
+): TransactionPromise<Proven, Signed> {
   const pending = getPromise().then();
   return Object.assign(pending, {
-    sign(...args: Parameters<Transaction['sign']>) {
-      return toTransactionPromise(() => pending.then((v) => v.sign(...args)));
+    sign(...args: Parameters<Transaction<boolean, false>['sign']>) {
+      return toTransactionPromise(() =>
+        pending.then((v) => (v as Transaction<Proven, false>).sign(...args))
+      );
     },
     send() {
       return toPendingTransactionPromise(() => pending.then((v) => v.send()));
     },
-  }) as TransactionPromise;
+    prove() {
+      return toTransactionPromise(() =>
+        pending.then((v) => (v as never as Transaction<false, Signed>).prove())
+      );
+    },
+    proofs: pending
+      .then((v) => (v as never as Transaction<true, Proven>).proofs)
+      .catch(
```

Comment:
@MartinMinkov and I were wondering whether this is acceptable?: if the transaction fails, then `proofs` resolves to an empty array.

This would solve the problem that was causing the `hello-world/run.ts` integration test to fail: the `pending.then` created a new promise, which escaped the `try`. Aka. the issue didn't have to do with the integration test, but rather an unsuppressed throw here.

Reply:
I don't think this would be an issue, as it's really the only case in which you have a promise that can't be contained by the caller's try/catch.